README.md
# PlaceGPT module for FoundryVTT
<p align="center">
<img alt="GitHub release (latest by date)" src="https://img.shields.io/github/v/release/megastruktur/foundryvtt-place-gpt"> <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/megastruktur/foundryvtt-place-gpt"> <img alt="GitHub All Releases" src="https://img.shields.io/github/downloads/megastruktur/foundryvtt-place-gpt/total" /> <img alt="GitHub Release Date" src="https://img.shields.io/github/release-date/megastruktur/foundryvtt-place-gpt?label=latest%20release" /> 
</p>
<p align="center">
<img alt="GitHub" src="https://img.shields.io/github/license/megastruktur/foundryvtt-place-gpt"> <a href="https://github.com/megastruktur/foundryvtt-place-gpt/issues"><img alt="GitHub issues" src="https://img.shields.io/github/issues/megastruktur/foundryvtt-place-gpt"></a> <a href="https://github.com/megastruktur/foundryvtt-place-gpt/network"><img alt="GitHub forks" src="https://img.shields.io/github/forks/megastruktur/foundryvtt-place-gpt"></a> <a href="https://github.com/megastruktur/foundryvtt-place-gpt/stargazers"><img alt="GitHub stars" src="https://img.shields.io/github/stars/megastruktur/foundryvtt-place-gpt"></a> 
</p>

# WARNING
**The module works only with ChatGPT for now so if you don't have openai account you won't be able to use it for now.**

## Description
This module allows you to generate places using the GPT model (currently ChatGPT3).
The Place is created as Journal entry by default but additionally can be added as Chat message (see module Settings).

![Launch](./screenshots/Launch.gif?raw=true "Launch")
![Result](./screenshots/Result.gif?raw=true "Result")


## Installation

### ChatGPT
0. Make sure to get familiar with their pricing first at https://openai.com/api/pricing/
1. Install and enable the module
2. Create account at openai.com (or use an existing one)
3. Go to https://platform.openai.com/account/api-keys and create a new API key
4. Go to module settings and insert the API key from the step above.

## Generators currently available
- ChatGPT (gpt-3.5-turbo-0301 model with temperature 0.2)

## To be done
1. Add more random generators (see the ./ai_classes directory)

CHANGELOG.txt
2.0
- Global refactoring
- Adds a new "Dummy" GPT provider with test data instead of a checkbox.
- Adds support for other languages generated by the GPT provider.

1.0
- The module is born.

place-gpt.js
// CONFIG.debug.hooks = true

import {AiManager} from "./ai-classes/ai-manager.js";

// @todo change the renderSceneControls to getSceneControlButtons
Hooks.on("renderSceneControls", async (app, html) => {
    let place_gpt = $('<li class="scene-control" title="Create Place"><i class="fas fa-globe"></i></li>');
    place_gpt.click( async function() {
        await PlaceGPT.generatePopup().render(true)
    });
    html.children().first().append( place_gpt );
});

// @todo WIP
// Hooks.on("renderJournalPageSheet", (sheet, view, data) => {
//     view.find("a.placegpt-generate-room-description").click( ev => {
//         PlaceGPT.clickGenerateDescription(ev, sheet.object.parent._id);
//     });
// });

Hooks.on("ready", () => {
    console.log("place-gpt | Initializing AI");
    game.placeGPTAI = new AiManager().load_ai();
    game.placeGPT = PlaceGPT;
    console.log("place-gpt | Ready with AI provider: " + game.placeGPTAI.name);
});

class PlaceGPT {

    static async waitingForGpt(place, full_or_sections = "section") {
        const dialog = PlaceGPT.waitingPopup();
        dialog.render(true);

        // While the call is being made, show the waiting popup.
        // When the call is done, close the waiting popup and show the response in chat.
        const promise = game.placeGPTAI.getCompletion(place, full_or_sections);

        promise.then(
            (data) => {
                dialog.close();
                if (data !== null) {
                    if (game.settings.get("place-gpt", "outputTo") === 0 || game.settings.get("place-gpt", "outputTo") === 2) {
                        PlaceGPT.journalEntryPlaceDescription(place, data);
                    }
                    if (game.settings.get("place-gpt", "outputTo") === 1 || game.settings.get("place-gpt", "outputTo") === 2) {
                        PlaceGPT.chatPlaceDescription(place, data);
                    }
                    ui.notifications.info(game.i18n.localize('place-gpt.success.success_generating_place'));
                }
                else {
                    ui.notifications.error(game.i18n.localize('place-gpt.error.error_generating_place'));
                }
            }
        )
            .catch(function(error) {
                dialog.close();
                console.log(error);
                ui.notifications.error(game.i18n.localize('place-gpt.error.error_generating_place'));
            });
    }


    /**
     * @todo WIP
     * @param event
     * @param journal_id
     */
    // static clickGenerateDescription(event, journal_id) {
    //     event.preventDefault();
    //     const journal = game.journal.get(journal_id);
    //     const page_id = event.target.closest("article.journal-entry-page").getAttribute("data-page-id");
    //     const page = journal.pages.get(page_id);
    //     let section_description = event.target.parentElement.children[0].innerText;
    //     let page_html = page.text.content;
    //     page.update({"text.content": page.text.content + "<div>Test</div>"});
    //     console.log(`place-gpt | Generating description for Journal "${journal.name}"`);
    // }


    /**
     *
     * Place example:
     * [{ "name": "Bridge", "description": "This is the command center of the spaceship. The walls are lined with screens displaying various readings and data. The captain's chair sits at the center of the room, facing the main viewscreen. There are doors leading to the rest of the ship.", "exits": { "north": "Captain's Quarters", "east": "Navigation Room", "south": "Main Corridor" }]
     *
     * @param placeDescription
     * @param placeJSON
     */
    static chatPlaceDescription(placeDescription, placeJSON) {

        // Go foreach placeJSON and create a text for each room:
        // RoomName
        // Description
        // Exits: North, East, South, West
        // Create a chat message with the text
        let placeGenerated = `<h1>${placeDescription}</h1>`;
        placeGenerated += PlaceGPT.parsePlaceJSONToHTML(placeDescription, placeJSON);

        ChatMessage.create({
            content: placeGenerated,
            speaker: ChatMessage.getSpeaker({ alias: placeDescription }),
        });

    }

    /**
     * @todo What if the entry exists already?
     * @param placeDescription
     * @param placeJSON
     */
    static journalEntryPlaceDescription(placeDescription, placeJSON) {
        JournalEntry.create({
            name: placeDescription,
            content: PlaceGPT.parsePlaceJSONToHTML(placeDescription, placeJSON),
        });
    }

    static parsePlaceJSONToHTML(placeDescription, placeJSON) {
        let placeGenerated = "";
        placeJSON.forEach(room => {
            placeGenerated += `<div class="room">`;
            placeGenerated += `<h2>${room.name}</h2>`;
            placeGenerated += `<p class="placegpt-section-description">`;

            if (room.description !== undefined) {
                placeGenerated += room.description;
            }
            else {
                placeGenerated += `<a class="placegpt-generate-room-description">${game.i18n.localize("place-gpt.generate-room-description")}</a>`;
            }

            placeGenerated += `</p>`;

            // check if room.exits exists
            if (room.exits !== undefined) {
                placeGenerated += `<h3>${game.i18n.localize("place-gpt.Exits")}</h3>`;
                let exits = "";
                for (let [key, value] of Object.entries(room.exits)) {
                    exits += `<p><b>${key}</b>: ${value}</p>`;
                }
                placeGenerated += exits;
            }
            placeGenerated += `</div>`;
        });
        return placeGenerated;
    }


    /**
     * Waiting Popup Dialog
     * @returns {Dialog}
     */
    static waitingPopup() {
        return new Dialog({
            title: `${game.i18n.localize("place-gpt.Generating")}`,
            content: `
      <h2>${game.i18n.localize("place-gpt.generating_the_place")}</h2>
      <form>
        <div className="form-group">
          <label>${game.i18n.localize("place-gpt.waiting_for_generate")}</label>
        </div><br/>
      </form>
    `,
            buttons: {
                cancel: {
                    icon: "<i class='fas fa-times'></i>",
                    label: game.i18n.localize('Cancel'),
                },
            },
            default: "close",
        });
    }


    /**
     * Generate Popup Dialog
     * @returns {Dialog}
     */
    static generatePopup() {
        return new Dialog({
            title: game.i18n.localize("place-gpt.generate_a_place"),
            content: `
      <h2>${game.i18n.localize("place-gpt.generate_a_place")}</h2>
      <form>
        <div className="form-group">
          <textarea id="place-gpt-place" name="place-gpt-place" value=""></textarea>
        </div>
        <br/>
      </form>
    `,
            buttons: {
                generate: {
                    icon: "<i class='fas fa-check'></i>",
                    label: game.i18n.localize("place-gpt.Generate"),
                    callback: (html) => {
                        // Get the text prompt from the form
                        let place = html.find('[name="place-gpt-place"]')[0].value;

                        // @todo section loader
                        let full_or_sections = "full";
                        if (html.find('[name="place-gpt-full_or_sections"]')[0] !== undefined) {
                            full_or_sections = html.find('[name="place-gpt-full_or_sections"]')[0].value;
                        }

                        if (place === "") {
                            throw new Error(game.i18n.localize('place-gpt.error.prompt_required'));
                        }
                        PlaceGPT.waitingForGpt(place, full_or_sections);
                    },
                },
                cancel: {
                    icon: "<i class='fas fa-times'></i>",
                    label: game.i18n.localize('Cancel'),
                },
            },
            default: "generate",
        });
    }
}


placegpt-build.sh
#!/bin/bash

# CLI parameter should be OPENAI_TOKEN
# If not, exit
if [ -z "$1" ]
  then
    echo "No OPENAI_TOKEN supplied"
    exit 1
fi

OPENAI_TOKEN=$1

# Read the CHANGELOG.txt and take the first line. Thi is the VERSION
VERSION=$(head -n 1 CHANGELOG.txt)

# Create a directory releases/$VERSION
# Copy all files and directories to releases/$VERSION except .git and releases
mkdir -p releases/"$VERSION"
rsync -av --exclude='.git/*' --exclude='releases/*' --exclude='.*' --exclude='placegpt-build.sh' . releases/$VERSION

# Search string OPENAITEMPDEMO in releases/$VERSION/ai-classes/chatgpt-temporary-demo.js
# Replace it with the value $OPENAI_TOKEN
echo "Replacing the token in releases/${VERSION}/ai-classes/chatgpt-temporary-demo.js"
sed -i "s/OPENAITEMPDEMO/${OPENAI_TOKEN}/g" "releases/${VERSION}/ai-classes/chatgpt-temporary-demo.js"


#zip -r releases/place-gpt-${VERSION}.zip place-gpt -x ".*" ".git/*" "releases/*"

echo "Creating zip file"
cd releases
zip -r place-gpt-${VERSION}.zip ${VERSION}

echo "Done"

scripts/config.js
import {ai_providers} from '../ai-classes/ai-manager.js';

Hooks.on('init', () => {
    game.settings.register('place-gpt', 'outputTo', {
        name: game.i18n.localize('place-gpt.settings.outputTo.name'),
        hint: game.i18n.localize('place-gpt.settings.outputTo.hint'),
        scope: 'world',
        config: true,
        type: Number,
        choices: {
            0: game.i18n.localize('place-gpt.settings.outputTo.journal'),
            1: game.i18n.localize('place-gpt.settings.outputTo.chat'),
            2: game.i18n.localize('place-gpt.settings.outputTo.both'),
        },
        default: 0,
    });

    // get the index and name for each AI provider
    // let ai_provider_choices = ["ChatGPT"];
    let ai_provider_choices = {};
    for (let [index, ai_provider] of Object.entries(ai_providers)) {
        ai_provider_choices[index] = ai_provider.name;
    }
    game.settings.register('place-gpt', 'ai_selector', {
        name: game.i18n.localize('place-gpt.settings.ai_selector.name'),
        hint: game.i18n.localize('place-gpt.settings.ai_selector.hint'),
        scope: 'world',
        config: true,
        type: Number,
        choices: ai_provider_choices,
        default: 0,
    });
});


ai-classes/dummy.js
export class Dummy {

    get name() {
        return 'Dummy';
    }

    /**
     * Get a response from ChatGPT
     * @param prompt
     * @param full_or_sections
     * @returns {Promise<any|null>}
     */
    async getCompletion(prompt, full_or_sections= "section") {

        console.log(full_or_sections);
        let apiUrl = '/modules/place-gpt/assets/dummy_section.json';
        if (full_or_sections === "full") {
            apiUrl = '/modules/place-gpt/assets/dummy_full.json';
        }

        const headers = {
            'Content-Type': 'application/json',
        };

        const requestOptions = {
            method: 'GET',
            headers: headers
        };

        try {

            await new Promise(r => setTimeout(r, 2000));

            const response = await fetch(apiUrl, requestOptions);
            const data = await response.json();

            // Retrieve the assistant's reply from the API response
            // Try to decode JSON (if the response is JSON)
            try {
                return data;
            }
            catch(error) {
                // If the response is not JSON, return the raw string
                console.error('place-gpt Error:', error);
                return null;
            }
        } catch (error) {
            console.error('place-gpt Error:', error);
            // Handle error
            return null;
        }
    }
}

ai-classes/chatgpt-temporary-demo.js
export class ChatGPTTD {

    get name() {
        return 'ChatGPT Temporary Demo';
    }

    /**
     * Get a response from ChatGPT
     * @param prompt
     * @param full_or_sections
     * @returns {Promise<any|null>}
     */
    async getCompletion(prompt, full_or_sections = "section") {

        if (!game.settings.get('place-gpt', 'openaiAPIToken')) {
            ui.notifications.error(game.i18n.localize('place-gpt.error.no_api_token'));
            return null;
        }

        const apiUrl = 'https://api.openai.com/v1/chat/completions';

        let full_or_sections_prompt = game.i18n.localize("place-gpt.prompt_section");
        if (full_or_sections === "full") {
            full_or_sections_prompt = game.i18n.localize("place-gpt.prompt_full")
        }

        const headers = {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer OPENAITEMPDEMO'
        };

        const body = {
            'model': 'gpt-3.5-turbo-0301',
            'temperature': 0.2,
            'messages': [
                {'role': 'system', 'content': full_or_sections_prompt},
                {'role': 'system', 'content': "Answer in the language of user request but keep the JSON valid."},
                {'role': 'user', 'content': prompt}
            ]
        };

        const requestOptions = {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(body)
        };

        try {
            // @todo add different http error handling e.g. 429 is out of credits.
            const response = await fetch(apiUrl, requestOptions);
            const data = await response.json();

            // Retrieve the assistant's reply from the API response
            // Try to decode JSON (if the response is JSON)
            try {
                return JSON.parse(data.choices[0].message.content);
            }
            catch {
                // If the response is not JSON, return the raw string
                return null;
            }
        } catch (error) {
            console.error('Error:', error);
            // Handle error
            return null;
        }
    }
}

ai-classes/chatgpt.js
Hooks.on('init', () => {
    game.settings.register('place-gpt', 'openaiAPIToken', {
        name: 'OpenAI API Token',
        hint: 'Get the token at the OpenAPI developers console',
        scope: 'world',
        config: true,
        type: String,
        default: ''
    });
});

export class ChatGPT {

    get name() {
        return 'ChatGPT';
    }

    /**
     * Get a response from ChatGPT
     * @param prompt
     * @param full_or_sections
     * @returns {Promise<any|null>}
     */
    async getCompletion(prompt, full_or_sections = "section") {

        if (!game.settings.get('place-gpt', 'openaiAPIToken')) {
            ui.notifications.error(game.i18n.localize('place-gpt.error.no_api_token'));
            return null;
        }

        const apiUrl = 'https://api.openai.com/v1/chat/completions';

        let full_or_sections_prompt = game.i18n.localize("place-gpt.prompt_section");
        if (full_or_sections === "full") {
            full_or_sections_prompt = game.i18n.localize("place-gpt.prompt_full")
        }

        const headers = {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + game.settings.get('place-gpt', 'openaiAPIToken')
        };

        const body = {
            'model': 'gpt-3.5-turbo-0301',
            'temperature': 0.2,
            'messages': [
                {'role': 'system', 'content': full_or_sections_prompt},
                {'role': 'system', 'content': "Answer in the language of user request but keep the JSON valid."},
                {'role': 'user', 'content': prompt}
            ]
        };

        const requestOptions = {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(body)
        };

        try {
            // @todo add different http error handling e.g. 429 is out of credits.
            const response = await fetch(apiUrl, requestOptions);
            const data = await response.json();

            // Retrieve the assistant's reply from the API response
            // Try to decode JSON (if the response is JSON)
            try {
                return JSON.parse(data.choices[0].message.content);
            }
            catch {
                // If the response is not JSON, return the raw string
                return null;
            }
        } catch (error) {
            console.error('Error:', error);
            // Handle error
            return null;
        }
    }
}